### ISSUE TITLE
connectionId and timestamp missing from presence members

### ISSUE THREAD

**Issue by mattheworiordan** [ably/ably-js#310]
See https://jsbin.ably.io/ebemoj/4/edit

Why are the members returned in the `get` method missing all the attributes of Presence member such as `connectionId`.  This is sorely needed for anyone trying to work out if they are present or not upon reconnection.

**Comment by SimonWoolf** [comment]
@mattheworiordan they're not, they're just hidden from the default `toJSON()`, see See https://jsbin.ably.io/ebemoj/6/ that uses `toString()` instead, see past discussion at https://github.com/ably/ably-js/issues/44

**Comment by SimonWoolf** [comment]
(FWIW I'd be all in favour of making internal library use use a custom method rather than toJSON, so end-users' JSON.stringify() calls aren't affected, but Paddy prefers this way. shrug)

**Comment by paddybyers** [comment]
The only reason for having a `toJSON()` is because that's how you control what gets packed when sending on a connection (both with JSON and msgpack transports). Since we have custom handling for `data` (ie base64-encoding) then we need to do this. I suspect the current `toJSON()` methods omit those fields because they're not required to be sent but I'm fine with adding them.

**Comment by SimonWoolf** [comment]
> that's how you control what gets packed when sending on a connection (both with JSON and msgpack transports). Since we have custom handling for data (ie base64-encoding) then we need to do this.

Do we really, though? Why can't we just do it explicitly, rather than overriding toJSON? Surely what fields we need to send to Ably shouldn't have to be linked to what the users sees with a stringify?

Code speaks louder than words, so quick-and-dirty proposal: https://github.com/ably/ably-js/compare/prepareforsending?expand=1

**Comment by paddybyers** [comment]
Of course the reason for using the `toJSON()` hack was to avoid having to write code not only for the types in question but also each container type that might embed them.

**Comment by mattheworiordan** [comment]
I like your change @SimonWoolf, overload `toJSON` is a bit hacky, it's the kind of thing I'd expect in my code not that of @paddybyers :)

BTW. I don't like the name `prepareForSending` given it's not preparing but instead bundling and generating a new payload, so perhaps `packageForSending` or `toAblyPayload()` etc?

**Comment by SimonWoolf** [comment]
TBH I don't feel very strongly about this; if Paddy believes there's value in the toJSON way, I'm ok with that. I'll revert https://github.com/ably/ably-js/pull/117/commits/c7081323613c2242c64c2024dccdbfbbf4dcbbe4

**Comment by mattheworiordan** [comment]
@paddybyers WDYT?

**Comment by paddybyers** [comment]
I agree `toJSON()`, whilst theoretically the most efficient, comes with some baggage and constraints, and it only just works because of the hack to detect whether it was invoked via `JSON.stringify()` or msgpack.

So I'm not opposed to changing it.

As for naming, in realtime for the same functionality we have `pack()` and `unpack()`, which would be fine, or `serialise()/deserialise()`, `transportEncode()` etc.

**Comment by mattheworiordan** [comment]
Is this covered in 0.9 now @SimonWoolf?

**Comment by SimonWoolf** [comment]
Nope, as of now it's still using the toJSON hack. (There was some issue with the `prepareforsending` approach, can't remember what now, that made it not mergeable as-is, then other things took priority). Would still be nice to refactor that at some point, so leaving this issue open. Don't think it's high priority though

**Comment by mattheworiordan** [comment]
> Don't think it's high priority though

Ok, thanks for clarifying

**Comment by ttypic** [comment]
No longer relevant
